x/concentrated-liquidity/math/math_test.go

```diff
@@ -150,25 +151,87 @@ func (suite *ConcentratedMathTestSuite) TestGetNextSqrtPriceFromAmount1RoundingD
+			isWithTolerance: false,
 			// https://www.wolframalpha.com/input?i=%281517882343.751510418088349649+*+%2874.161984870956629487+-+70.710678118654752440+%29%29+%2F+%2870.710678118654752440+*+74.161984870956629487%29
 		},
+		"round down: large liquidity amount in wide price range": {
+			// Note the values are hand-picked to cause multiplication of 2 large numbers
+			// causing the magnitude of truncations to be larger
+			// while staying under bit length of sdk.Dec
+			// from decimal import *
+			// from math import *
+			// getcontext().prec = 100
+			// max_sqrt_p = Decimal("30860351331.852813530648276680")
+			// min_sqrt_p = Decimal("0.000000152731791058")
+			// liq = Decimal("931361973132462178951297")
+			// liq * (max_sqrt_p - min_sqrt_p) / (max_sqrt_p * min_sqrt_p)
+			liquidity: sdk.MustNewDecFromStr("931361973132462178951297"),
+			// price: 0.000000000000023327
+			sqrtPA: sdk.MustNewDecFromStr("0.000000152731791058"),
+			// price: 952361284325389721913
+			sqrtPB:          sdk.MustNewDecFromStr("30860351331.852813530648276680"),
+			roundUp:         false,
+			amount0Expected: sdk.MustNewDecFromStr("6098022989717817431593106314408.888128101590393209").String(), // truncated at precision end.
+			isWithTolerance: true,
+		},
+		"round up: large liquidity amount in wide price range": {
+			// Note the values are hand-picked to cause multiplication of 2 large numbers
+			// causing the magnitude of truncations to be larger
+			// while staying under bit length of sdk.Dec
+			// from decimal import *
+			// from math import *
+			// getcontext().prec = 100
+			// max_sqrt_p = Decimal("30860351331.852813530648276680")
+			// min_sqrt_p = Decimal("0.000000152731791058")
+			// liq = Decimal("931361973132462178951297")
+			// liq * (max_sqrt_p - min_sqrt_p) / (max_sqrt_p * min_sqrt_p)
+			liquidity: sdk.MustNewDecFromStr("931361973132462178951297"),
+			// price: 0.000000000000023327
+			sqrtPA: sdk.MustNewDecFromStr("0.000000152731791058"),
+			// price: 952361284325389721913
+			sqrtPB:          sdk.MustNewDecFromStr("30860351331.852813530648276680"),
+			roundUp:         true,
+			amount0Expected: sdk.MustNewDecFromStr("6098022989717817431593106314408.888128101590393209").Ceil().String(), // truncated at precision end.
+			isWithTolerance: true,
+		},
 	}
 
 	for name, tc := range testCases {
 		tc := tc
 
 		suite.Run(name, func() {
-			amount0 := math.CalcAmount0Delta(tc.liquidity, tc.sqrtPCurrent, tc.sqrtPUpper, false)
-			suite.Require().Equal(tc.amount0Expected, amount0.String())
+			amount0 := math.CalcAmount0Delta(tc.liquidity, tc.sqrtPA, tc.sqrtPB, tc.roundUp)
+
+			if !tc.isWithTolerance {
+				suite.Require().Equal(tc.amount0Expected, amount0.String())
+				return
+			}
+
+			roundingDir := osmomath.RoundUp
+			if !tc.roundUp {
+				roundingDir = osmomath.RoundDown
+			}
+
+			tolerance := osmomath.ErrTolerance{
+				MultiplicativeTolerance: sdk.SmallestDec(),
+				RoundingDir:             roundingDir,
+			}
```

Comment:
Note: this invariant would not hold with the old rounding logic.

The additive error is in the order of magnitude of 6 with these numbers.